### PR TITLE
keystore: add a basic in-memory keystore

### PR DESCRIPTION
**Pull request by onyb**
### What is this?

|JIRA| [`BACK-757`](https://ledgerhq.atlassian.net/browse/BACK-757) | [`BACK-758`](https://ledgerhq.atlassian.net/browse/BACK-758)
-|-|-

This PR adds an implementation of the `Keystore` interface called `InMemoryKeystore`. It also includes a simple parser to extract relevant information from an output descriptor string, which is also used as a **Keychain URI**.

### CODE REVIEW

File: bitcoin/service.proto [hunk 1]
```diff
@@ -0,0 +1,150 @@
+syntax = "proto3";
+
+package bitcoin;
+option go_package = ".;bitcoin";
+option java_package = "co.ledger.protobuf";
```
Comment (vbergeron-ledger) [inline bitcoin/service.proto:5]:
```suggestion
option java_package = "co.ledger.keychain.protobuf";
```
Add a project / service identifier to package.

File: bitcoin/service.proto [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+//
+// This enum type may be used by gRPC clients to differentiate protocol
+// behaviour, magic numbers, addresses, keys, etc., for one network from those
+// intended for use on another network.
+enum BitcoinNetwork {
+  BITCOIN_NETWORK_UNSPECIFIED = 0;  // Fallback value if unrecognized / unspecified
+  BITCOIN_NETWORK_MAINNET     = 1;  // Main network
+  BITCOIN_NETWORK_TESTNET3    = 2;  // Current test network (since Bitcoin Core v0.7)
+  BITCOIN_NETWORK_REGTEST     = 3;  // Regression test network
+}
```
Comment (vbergeron-ledger) [inline bitcoin/service.proto:47]: Why do we have to handle the unspecified case ?

File: bitcoin/service.proto [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+message ValidateAddressResponse {
+  // Address in normalized form, if valid; original address otherwise.
+  string address = 1;
+
+  // Whether the input address is valid or not.
+  bool is_valid = 2;
+
+  // Human-readable reason for the address being invalid. Use ONLY if is_valid
+  // is false.
+  string invalid_reason = 3;
```
Comment (vbergeron-ledger) [inline bitcoin/service.proto:83]: What do you think of :
```
message InvalidAddress {
   string reason
}
message ValidAddress {
  string value
}
oneof {
 ValidAddress address
 InvalidAddress error
}
```

File: pkg/keystore/descriptor.go [hunk 4]
```diff
@@ -1 +1,53 @@
...
+	var scheme Scheme
+
+	if strings.HasPrefix(descriptor, "sh(wpkh(") {
+		scheme = BIP49
+	} else if strings.HasPrefix(descriptor, "wpkh(") {
+		scheme = BIP84
+	} else if strings.HasPrefix(descriptor, "pkh(") {
+		scheme = BIP44
+	} else {
+		return DescriptorTokens{}, errors.New("unrecognized scheme")
```
Comment (vbergeron-ledger) [inline pkg/keystore/descriptor.go:33]: Proposal : include the non recognized scheme in the error message.

File: pkg/keystore/descriptor.go [hunk 5]
```diff
@@ -1 +1,53 @@
...
+	}
+
+	// Match the base key using regex, and ignore everything else.
+	// FIXME: make the regex match more robust.
+	r := regexp.MustCompile(`.*\((?:\[.*])?([\w+]*).*\).*`)
```
Comment (vbergeron-ledger) [inline pkg/keystore/descriptor.go:38]: no need compiling the regexp each time you parse a descriptor.

File: pkg/keystore/descriptor.go [hunk 6]
```diff
@@ -1 +1,53 @@
...
+	groups := r.FindStringSubmatch(descriptor)
+	if len(groups) != 2 {
+		return DescriptorTokens{}, errors.New("invalid descriptor")
+	}
+
+	xpub := groups[1]
+	if xpub == "" {
+		return DescriptorTokens{}, errors.New("invalid descriptor")
+	}
```
Comment (vbergeron-ledger) [inline pkg/keystore/descriptor.go:47]: Proposal : include the faulty input to the error message.

File: pkg/keystore/memory.go [hunk 7]
```diff
@@ -21,18 +40,59 @@ func (s *InMemoryKeystore) Get(descriptor string) (KeychainInfo, error) {
...
+		child, err := s.client.DeriveExtendedKey(
+			context.Background(), &bitcoin.DeriveExtendedKeyRequest{
+				ExtendedKey: tokens.XPub,
+				Derivation:  []uint32{childIndex},
+			})
+		if err != nil {
+			return "", "", nil
+		}
+
+		return hex.EncodeToString(child.PublicKey), hex.EncodeToString(child.ChainCode), nil
```
Comment (vbergeron-ledger) [inline pkg/keystore/memory.go:64]: If I understand correctly this is a closure right ? Why do we need one there ?

**Review by vbergeron-ledger — APPROVED** [review]

File: pkg/keystore/memory.go (hunk 7 above)
Comment (onyb) [reply pkg/keystore/memory.go:64]: Yes it's a closure. I wanted to reuse the logic for deriving the pubkey/chaincode for internal and external chain paths. I could've used an isolated helper actually.

File: bitcoin/service.proto (hunk 1 above)
Comment (onyb) [reply bitcoin/service.proto:5]: This proto is not the one for keychain service, but rather for bitcoin-svc, so I guess I'll use `co.ledger.bitcoin.protobuf` here.

Anyway, on the client-side you can change the package to fit your desired directory structure.

File: bitcoin/service.proto (hunk 2 above)
Comment (onyb) [reply bitcoin/service.proto:47]: It's basically a fallback if a client tries to use a network that has been removed by the server. Also recommended by [`buf.build`](https://buf.build/docs/introduction) lint checks.

File: bitcoin/service.proto (hunk 3 above)
Comment (onyb) [reply bitcoin/service.proto:83]: I like it! Will change it in the bitcoin-svc repo. 👍

File: pkg/keystore/descriptor.go (hunk 4 above)
Comment (onyb) [reply pkg/keystore/descriptor.go]: Fixed all error related comments, so going to mark the conversations as resolved.

File: pkg/keystore/descriptor.go (hunk 6 above)
Comment (onyb) [reply pkg/keystore/descriptor.go]: Thanks. I fixed it with wrapped errors using https://github.com/pkg/errors.

**Comment by onyb** [comment]
I'm going to merge this PR as-is, as the suggested changes are to be done in `bitcoin-svc` repo first.

**Pull request merged**